### PR TITLE
Add an endpoint for updating customer by order

### LINKED ISSUE #9461: Can't update customer information on the pay for order page
## Is your feature request related to a problem? Please describe.
On the pay-for-order page, if a shopper updates their information, the checkout page will render an empty cart error.

## Describe the solution you'd like
Add a new `OrderUpdateCustomer` endpoint.  If it's a pay-for-order, in the redux store action `updateCustomerData`, we call the `OrderUpdateCustomer` endpoint instead of `/wc/store/v1/cart/update-customer` 

## Describe alternatives you've considered
When calling `updateCustomerData`, we can call the order's endpoint again instead of receiving cart items from the update customer endpoint.

## Recording

https://github.com/woocommerce/woocommerce-blocks/assets/56378160/20af238f-6b46-4ace-8617-9fc503bc0101

### PR DESCRIPTION
**Pull request by hsingyuc**
Fixes #9461 

#### Other Checks

- [x] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

Screen Shot 2023-07-09 at 6 42 20 PM

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### Testing instruction
* Test checkout as a guest*
1. Install and activate a plugin with pay-for-order flow. ex WooCommerce-Bookings
2. Create a product that needs confirmation
3. Purchase the product with pay-for-order flow
1. Go to WooCommerce -> Orders
1. Make sure the Customer is `Guest` 
1. Click the `Customer payment page` to go to the pay-for-order page
1. Copy the `order_id` and the `key` from the URL
1. Use Postman to test this endpoint http://merchant.local/wp-json/wc/store/order/YOUR_ORDER_NUMBER/update-customer?key=YOUR_KEY&billing_email=YOUR_BILLING_EMAIL
2. Make sure on Postman the authorization type set to `No Auth` 
3. POST to update billing & shipping address. Ex. last name
```
{
  "billing_address": {
    "last_name": "Guest"
  },
  "shipping_address": {
    "last_name": "Guest"
  }
}
```
11. Check that the billing address and the shipping address are updated

*Test checkout as an existing customer*
1. Use the same order you created above
9. Change the Customer to an existing customer
3. Go to WooCommerce -> Settings -> Advanced -> Rest API
4. Add key with the customer email, read/write permission, copy the Consumer key and secret
5. Go to the same request on Postman, change the authorization type to `OAuth 1.0`, and paste the key and secret there
6. Remove the `key` and `billing_email` params
7. Update the POST body to a new last name so you can see the update later 
8. Send the request
9. Go back to WooCommerce -> Orders -> your order
10. Check the billing and shipping last names are updated

* [ ] Do not include in the Testing Notes 

### Changelog

> Add an endpoint for updating customers by order

### CODE REVIEW

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (leonardola) [inline src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Do we need to run the sanitization twice?

File: src/StoreApi/Utilities/UpdateCustomerTrait.php
Comment (leonardola) [inline src/StoreApi/Utilities/UpdateCustomerTrait.php:116]: Is this still needed?

**Review by leonardola — APPROVED** [review]
Code looks good and test went well. Left a few questions.

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Good call! Removed here [469b730](https://github.com/woocommerce/woocommerce-blocks/pull/9743/commits/469b7304715341d30d75b7d5b6629f089c08399d).

File: src/StoreApi/Utilities/UpdateCustomerTrait.php
Comment (hsingyuc) [reply src/StoreApi/Utilities/UpdateCustomerTrait.php:116]: @senadir This part was introduced in [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8460) and I can't find more information about it.  Do you recall or know if it's fixed?

**Comment by hsingyuc** [comment]
@leonardola  Thank you for reviewing the PR promptly!

**Comment by hsingyuc** [comment]
@mikejolley and @wavvves  This PR is ready for review.
@senadir Could you take a look at [this](https://github.com/woocommerce/woocommerce-blocks/pull/9743#discussion_r1258695121) when you have time?

**Comment by wavvves** [comment]
Following the testing steps, I consistently get the following:

```
{
    "code": "woocommerce_rest_invalid_user",
    "message": "This order belongs to a different customer. Please log in to the correct account.",
    "data": {
        "status": 403
    }
}
```

Even though I am supplying the correct key and email.
Supplying Cart-Token and Nonce to try and authenticate has the same result.
I tried var dumping `wp_get_current_user()` just before the error is thrown to compare the supplied email, but no user is returned:

```
object(WP_User)#491 (8) {
  ["data"]=>
  object(stdClass)#495 (0) {
  }
  ["ID"]=>
  int(0)
  ["caps"]=>
  array(0) {
  }
  ["cap_key"]=>
  NULL
  ["roles"]=>
  array(0) {
  }
  ["allcaps"]=>
  array(0) {
  }
  ["filter"]=>
  NULL
  ["site_id":"WP_User":private]=>
  int(0)
}
```

Does anyone get the same results? @mikejolley ? @leonardola, what am I missing here from the testing steps?

As a side note, sending an `OPTIONS` request into this endpoint results in a `500 Internal Error`.

@hsingyuc if this is not going behind an experimental flag, then we should probably add some testing 🤔

**Comment by hsingyuc** [comment]
@wavvves  Thank you for reviewing the PR.  Maybe the testing instructions are not clear enough.  I've updated it, please let me know if any questions.

> if this is not going behind an experimental flag, then we should probably add some testing 

It's behind an experimental flag like order endpoint.

File: src/StoreApi/Utilities/OrderAuthorizationTrait.php
Comment (opr) [inline src/StoreApi/Utilities/OrderAuthorizationTrait.php]: Should we improve the name of this method? `validate_billing_email` doesn't describe very well what the method does.

I suggest `check_billing_email_matches_order` or similar.

File: src/StoreApi/Schemas/V1/OrderSchema.php
Comment (opr) [inline src/StoreApi/Schemas/V1/OrderSchema.php:274]: Nice, was this done because the order might have a different address to the one stored on the customer?

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [inline src/StoreApi/Routes/V1/OrderUpdateCustomer.php:97]: Sorry to be pedantic but can we call these `$billing_address` and `$shipping_address` just to make it a little easier to reason with?

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [inline src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Is there supposed to be a placeholder here like `x.x.x` or something so it updates when we do release this feature? Not sure how that works or how to ensure we track this to document it when it's released 🤔  any ideas?

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [inline src/StoreApi/Routes/V1/OrderUpdateCustomer.php:72]: Could you add a comment explaining this loop and why we're doing it? I get it but a comment would make it much quicker to understand when first coming to the code!

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [inline src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: It's a little unclear what the comment means exactly, is it saying the taxes _should_ be recalculated separately after using this, or that they will be recalculated as a result of running this? Please could you try to make that a little clearer? 

If the latter, I'm curious about the word `should`, that implies there's a case where they might not, is that true?

**Review by opr — CHANGES_REQUESTED** [review]
@hsingyuc kudos for the nice PR and the excellent testing instructions. I followed them and the PR works perfectly as described.

Just a couple of comments about code commenting really, please see if you agree and feel free to ping for a re-review when you've changed. 

Thanks!

File: src/StoreApi/Utilities/OrderAuthorizationTrait.php
Comment (hsingyuc) [reply src/StoreApi/Utilities/OrderAuthorizationTrait.php]: I'm not sure that this naming makes more sense to me.  I think this falls under validation and the word "check" is a little more ambiguous.

This function was already added before this PR, but if you think we need to rename it here, we could call this `validate_billing_email_matches_order` if that's more clear?

File: src/StoreApi/Schemas/V1/OrderSchema.php
Comment (hsingyuc) [reply src/StoreApi/Schemas/V1/OrderSchema.php:274]: This is because an order does not necessarily have a customer.  This will also get guest details.

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php:97]: Sure, we can update these.  For reference, these were following [existing patterns in the card endpoint](https://github.com/woocommerce/woocommerce-blocks/blob/469b7304715341d30d75b7d5b6629f089c08399d/src/StoreApi/Routes/V1/CartUpdateCustomer.php#L43).

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Good catch!  I'm not sure how this works either.  Do you know @mikejolley?

I only see one other `x.x.x` but it was 22 months ago and did not get replaced.

https://github.com/woocommerce/woocommerce-blocks/blob/469b7304715341d30d75b7d5b6629f089c08399d/src/InboxNotifications.php#L11

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: This is also copied from [the existing cart endpoint](https://github.com/woocommerce/woocommerce-blocks/blob/469b7304715341d30d75b7d5b6629f089c08399d/src/StoreApi/Routes/V1/CartUpdateCustomer.php#L10).

I'm happy to update this, but should we update both?

File: src/StoreApi/Utilities/OrderAuthorizationTrait.php
Comment (opr) [reply src/StoreApi/Utilities/OrderAuthorizationTrait.php]: > we could call this `validate_billing_email_matches_order` if that's more clear?

that would be perfect!

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php:97]: Ah, if the pattern already exists lets keep it as it is. No point diverging.

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (opr) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: That would be super helpful if you could update both, it adds a little noise to the PR but I think it's worth it.

File: src/StoreApi/Utilities/OrderAuthorizationTrait.php
Comment (hsingyuc) [reply src/StoreApi/Utilities/OrderAuthorizationTrait.php]: Updated in [a60597e](https://github.com/woocommerce/woocommerce-blocks/pull/9743/commits/a60597e0b8684a221ae4d7fe6e383901cdce0f2d).

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php:72]: Updated in [49d54f3](https://github.com/woocommerce/woocommerce-blocks/pull/9743/commits/49d54f337187530c649c360be3aec17a12cc5fb8).

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (hsingyuc) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Updated in [8934758](https://github.com/woocommerce/woocommerce-blocks/pull/9743/commits/8934758e5be61269c880ec677f893cb7b1cfa9e7).

**Review by opr — APPROVED** [review]
Perfect thanks @hsingyuc - I am approving on the basis that it works and all my feedback has been addressed.

@mikejolley do you know about the [`@since x.x.x` stuff](https://github.com/woocommerce/woocommerce-blocks/pull/9743/files#r1262505979)?

I'd wait for a second review from @wavvves before merging though. Thanks!

File: src/StoreApi/Routes/V1/OrderUpdateCustomer.php
Comment (nerrad) [reply src/StoreApi/Routes/V1/OrderUpdateCustomer.php]: Should be `@since $VID:$` and our release script will insert the correct version.

**Comment by wavvves** [comment]
Found a few issues. Please see p1689347400050409-slack-C8X6Q7XQU

**Comment by mikejolley** [comment]
> @mikejolley do you know about the [@since x.x.x stuff](https://github.com/woocommerce/woocommerce-blocks/pull/9743/files#r1262505979)?

Darren replied inline. There is a replacement syntax available.

**Comment by hsingyuc** [comment]
> Circumventing shipping costs

I spoke to Mike about a customer being able to update the shipping address and we agreed to allow customers to update shipping and billing.  How do you manage to place the order?  While you should be able to update the order, I think validation would prevent you from checking out with the order.

> Following the steps, I experimented with submitting invalid payloads and it failed big time with my first try. I submitted an object as the customer first_name:

Good catch! This sounds like we need sanitization for this property.

**Comment by hsingyuc** [comment]
>Sending an OPTIONS request to this endpoint results in a 500 error

I tested it and it works fine for me.  Did you add OAuth to Postman?

Screen Shot 2023-07-14 at 8 07 12 AM

**Comment by thealexandrelara** [comment]
I'm bumping the milestone for this PR to 10.8.0.

**Comment by wavvves** [comment]
> > Sending an OPTIONS request to this endpoint results in a 500 error
> 
> I tested it and it works fine for me. Did you add OAuth to Postman?

@hsingyuc I found the reason, it will throw a fatal on PHP 8 and just a warning on 7.4 , are you using PHP 7.4?

### PHP 7.4
```
PHP Warning	call_user_func() expects parameter 1 to be a valid callback, array must have exactly two members
WordPress core
File: [/wp-includes/rest-api/class-wp-rest-server.php](https://github.com/WordPress/wordpress-develop/blob/6.2.2/src/wp-includes/rest-api/class-wp-rest-server.php)
Line: 1497
```

### PHP 8
````
PHP Fatal	Uncaught TypeError: call_user_func(): Argument #1 ($callback) must be a valid callback, array must have exactly two members
Stack trace:
#0 /wp-includes/rest-api.php(785): WP_REST_Server->get_data_for_route('/wc/store/order...', Array, 'help')
#1 /wp-includes/class-wp-hook.php(308): rest_handle_options_request(Object(WP_REST_Response), Object(WP_REST_Server), Object(WP_REST_Request))
#2 /wp-includes/plugin.php(205): WP_Hook->apply_filters(NULL, Array)
#3 [/wp-includes/rest-api/class-wp-rest-server.php](https://github.com/WordPress/wordpress-develop/blob/6.2.2/src/wp-includes/rest-api/class-wp-rest-server.php)(984): apply_filters('rest_pre_dispat...', NULL, Object(WP_REST_Server), Object(WP_REST_Request))
#4 /wp-includes/rest-api/class-wp-rest-server.php(442): WP_REST_Server->dispatch(Object(WP_REST_Request))
#5 /wp-includes/rest-api.php(410): WP_REST_Server->serve_request('/wc/store/order...')
#6 /wp-includes/class-wp-hook.php(308): rest_api_loaded(Object(WP))
#7 /wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#8 /wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#9 /wp-includes/class-wp.php(399): do_action_ref_array('parse_request', Array)
#10 /wp-includes/class-wp.php(780): WP->parse_request('')
#11 /wp-includes/functions.php(1334): WP->main('')
#12 /wp-blog-header.php(16): wp()
#13 /index.php(17): require('/w...')
#14 {main} thrown in /wp-includes/rest-api/class-wp-rest-server.php on line 1497
WordPress core
File: /wp-includes/rest-api/class-wp-rest-server.php
Line: 1497
````

**Comment by wavvves** [comment]
>Good catch! This sounds like we need sanitization for this property.

@hsingyuc This is fixed in trunk now, just a matter of updating the PR 😉

**Comment by hsingyuc** [comment]
Thank you all for reviewing this PR.  I'm closing this PR based on our discussion here p1689368341767149-slack-C04NWH6NDHA

**Pull request closed without merging**
